The hidden keyword __makeref does not produce an error if used when System.TypedReference is not defined

If the code snippet below is compiled in a release build, the expected CS0518 error is not produced. I realize the the mkrefany opcode does not require the IL to refer to System.TypedReference, but any use of mkrefany without System.TypeReference is useless and should produce an error.

* Make a new Universal Windows Blank app 
* Drop this function in the project:
```
    public static T Read<T>() where T : new()
    {
        T result = new T();
        var refresult = __makeref(result);

        // ... method body

        return result;
    }
```
* Build under debug, you should get this error: error CS0518: Predefined type 'System.TypedReference' is not defined or imported
* Build under release and the C# compiler does not produce an error. (The .NET Native toolchain will produce an error, that I'm currently working to improve, but the C# compiler has the by far best capability to produce a good error.)

--- a/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenMscorlib.cs
+++ b/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenMscorlib.cs
@@ -252,6 +252,7 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
             var verifier = CompileAndVerify(compilation2);
         }
 
+        [Fact, WorkItem(3593, "https://github.com/dotnet/roslyn/issues/3593")]
         public void NoTypedRef()
         {
             var source1 =
@@ -285,6 +286,85 @@ public class C1
     // (7,25): error CS0518: Predefined type 'System.TypedReference' is not defined or imported
     //         var refresult = __makeref(result);
     Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "__makeref(result)").WithArguments("System.TypedReference").WithLocation(7, 25)
+);
+        }
+
+        [Fact, WorkItem(3746, "https://github.com/dotnet/roslyn/issues/3746")]
+        public void NoTypedRefBox()
+        {
+            var source1 =
+@"namespace System
+{
+    public class Object { }
+    public struct Void { }
+    public class ValueType { }
+    public struct Int32 { }
+    public struct Decimal { }
+}";
+            var compilation1 = CreateCompilation(source1, assemblyName: GetUniqueName());
+            var reference1 = MetadataReference.CreateFromStream(compilation1.EmitToStream());
+            var source2 =
+@"    
+public class C1
+{
+    public static object rrr;
+
+    public static T Read<T>() where T : new()
+    {
+        T result = new T();
+        var refresult = __makeref(result);
+        rrr = refresult;
+        return result;
+    }
+}
+";
+            var compilation2 = CreateCompilation(source2, new[] { reference1 });
+            compilation2.VerifyDiagnostics(
+    // (9,25): error CS0518: Predefined type 'System.TypedReference' is not defined or imported
+    //         var refresult = __makeref(result);
+    Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "__makeref(result)").WithArguments("System.TypedReference").WithLocation(9, 25)
+);
+        }
+
+        [Fact, WorkItem(3746, "https://github.com/dotnet/roslyn/issues/3746")]
+        public void NoTypedRefBox1()
+        {
+            var source1 =
+@"namespace System
+{
+    public class Object { }
+    public struct Void { }
+    public class ValueType { }
+    public struct Int32 { }
+    public struct Decimal { }
+    public struct TypedReference { }
+}";
+            var compilation1 = CreateCompilation(source1, assemblyName: GetUniqueName());
+            var reference1 = MetadataReference.CreateFromStream(compilation1.EmitToStream());
+            var source2 =
+@"    
+public class C1
+{
+    public static object rrr;
+
+    public static T Read<T>() where T : new()
+    {
+        T result = new T();
+        var refresult = __makeref(result);
+        rrr = refresult;
+        rrr = (object)__makeref(result);
+        return result;
+    }
+}
+";
+            var compilation2 = CreateCompilation(source2, new[] { reference1 });
+            compilation2.VerifyDiagnostics(
+    // (10,15): error CS0029: Cannot implicitly convert type 'System.TypedReference' to 'object'
+    //         rrr = refresult;
+    Diagnostic(ErrorCode.ERR_NoImplicitConv, "refresult").WithArguments("System.TypedReference", "object").WithLocation(10, 15),
+    // (11,15): error CS0030: Cannot convert type 'System.TypedReference' to 'object'
+    //         rrr = (object)__makeref(result);
+    Diagnostic(ErrorCode.ERR_NoExplicitConv, "(object)__makeref(result)").WithArguments("System.TypedReference", "object").WithLocation(11, 15)
 );
         }
 

